### ISSUE TITLE
Insanely Great Contributing Experience

### ISSUE THREAD

**Issue by ostera** [tldr-pages/tldr#1073]
We've already made learning commands easy. Our pages are a joy to read, concise yet friendly, with a fine balance between general and specific cases. And we keep working on making them better every day.

We've made finding commands very straightforward too. Although there obviously is room to make it even easier but that would mean pushing the clients forward with features like [tldr.jsx's Command Suggestion](https://github.com/ostera/tldr.jsx/issues/60) or [full text search](https://github.com/tldr-pages/tldr/issues/273), or even [Platform Detection](https://github.com/ostera/tldr.jsx/issues/70).

But there's something we're missing out and we are missing out _big time_, and that is lowering the bar for creating new pages to the point were creating an issue to request the page is actually _more work_ than proposing the page itself.

Here is a small proposal towards an insanely great contributing experience.

## Current UX

At the moment, adding a new command will likely begin with looking for it and not finding it, and from there on the roads are:

1. Submit an Command Request (Issue)
2. Submit a Command Proposal (PR)

Where the first one is fairly straightforward, and the second one can require a few steps, depending if the command already exists or not.

#### For an existing command

1. Find the source for the page you want to edit on Github
2. Click on `Edit`
3. Make your changes
4. Submit a Pull Request
5. Deal with feedback from Travis-CI and Contributors locally and update branch

#### For a new command

1. Find the platform for the page you want to add on Github
2. Click on `Create New File`
3. Write the page from scratch
4. Submit a Pull Request
5. Deal with feedback from Travis-CI and Contributors locally and update branch

And those are scenarios using Github's UI, which is the common denominator between all our users since many might not know how to use git well enough to go through the forking of the repo, cloning it locally, editing the file with whatever you want, going to Github and submitting the PR. Or using something like `hub`. 

## Proposed UX

Upon not finding a command, _if there is an Issue or PR open for that given command_, the clients will prompt you to chime in with the Issue or PR link readily available. This way being a part of the conversation becomes trivial.

If there is _no open issue or PR_, then a link to submit a proposal for the command should be readily available. Think `https://tldr.sh/macos/swift/edit`.

This will load a tldr-branded, web-based split markdown editor/documentation that

* has our command template by default
* checks the syntax and timely gives feedback on it as you edit
* shows the documentation for the given command if available (most posix/unix/gnu/bsd/osx commands have [man pages available online](https://github.com/tldr-pages/tldr/issues/1070)

An editor like this would already collapse the first 3 steps for adding a new command into a single one. Plus, since it doesn't require to remember the page format as it's suggested by the template and enforced by the linter, the review process will focus only on actual content.

After editing the page, a simple `Preview` switch will toggle a live preview as it would be shown on the web client. Any further changes can be made there until all syntax checks pass and you're comfortable with what you see.

A `Propose Command` button will then take you the PR for this command proposal.

#### Implementation Details

To get there we need to authenticate with Github via OAuth, and (links to API docs):

1. [Fork `tldr`](https://developer.github.com/v3/repos/forks/#create-a-fork)
2. [Create a file in a new branch*](https://developer.github.com/v3/repos/contents/#create-a-file)
3. [Create a PR](https://developer.github.com/v3/pulls/#create-a-pull-request)
4. Redirect to the PR.

\* this would work as well for updating a file, but it'd require one extra call to see if the file exists first, get it's current SHA1, and then make the same HTTP call adding the `sha` to the request body.

It feels natural to extend `tldr.jsx` to provide this functionality.

## Round-up

With the proposed UX, we went from finding out there's no command to submitting a command proposal in _a single step_: the guided edition of the command page on the tldr editor.

Thoughts? Comments?

**Comment by waldyrious** [comment]
I agree 100% with everything you say! Let's make a checklist of things to be done to get there :)

Just one question: I didn't understand the "locally" in "Deal with feedback from Travis-CI and Contributors locally and update branch". I thought you were describing a github-web-ui-based workflow?

**Comment by waldyrious** [comment]
Just [came across](https://tailordev.fr/blog/2016/10/13/communication-with-github-issues-2/) this snippet, thought it could be useful:

> We chose not to use [GitHub templates](https://github.com/blog/2111-issue-and-pull-request-templates) because, at the time of writing, it was not possible to declare multiple templates. Instead, we declare templates through preconfigured links (e.g., `https://github.com/<org>/<repo>/issues/new?body=<markdown>`) hidden behind image buttons in README files:
> ![](https://tailordev.fr/img/post/2016/10/preconfigured-issue.gif)

**Comment by waldyrious** [comment]
To make this issue more actionable, I changed the title, and will summarize the gist of the proposal here. @ostera is suggesting a custom web-based editor, tightly integrated with GitHub (via OAuth), that will allow a smoother way to propose new commands. In detail:

> Upon not finding a command, [...] a link to submit a proposal for the command should be readily available. Think `https://tldr.sh/macos/swift/edit`.
> 
> This will load a tldr-branded, web-based split markdown editor/documentation that
> - has our command template by default
> - checks the syntax and timely gives feedback on it as you edit
> - shows the documentation for the given command if available (most posix/unix/gnu/bsd/osx commands have [man pages available online](https://github.com/tldr-pages/tldr/issues/1070))
> 
> [...] A `Propose Command` button will then [create a] PR for this command proposal [and navigate to it].

I'll add the proposed behavior for linking to an existing issue requesting a command (or to the new issue page, if such an issue does not exist) to #1065, to keep this one focused.

**Comment by sbrl** [comment]
That would be cool, and might also simplify matters in #1065 a bit.

**Comment by Managor** [comment]
@spageektti

**Comment by spageektti** [comment]
Hi @managor 
Why you mentioned me?

**Comment by Managor** [comment]
You were working on somerhing like this. Jusr brought to your attention that this issue exists

**Comment by sbrl** [comment]
I'd forgotten about this one. It would be cool! I guess ti would be easiest as a desktop app, as doing complex git operations purely in the browser would be a difficult time.

**Comment by spageektti** [comment]
> You were working on somerhing like this. Just brought to your attention that this issue exists

Thanks! I created only the UI in PyQt. I am not working on this project currently because I have more important projects, contests, and school, but I plan to finish the app before June.
